### PR TITLE
#1476 :- Generate renditions for all images uploaded to magento

### LINKED ISSUE #1476: Generate renditions for all images uploaded to magento
### Description
Currently when the image is uploaded to the media gallery, the thumbnail is created for the image. The idea is to use a similar approach for generating the optimized version of the image.
 - When image is uploaded to the media gallery the optimized version of the image should be created ("Rendition")
 - Rendition should be created based on width/height specified in the configuration of MediaGalleryRenditions module
 - If the image width is greater than the width specified in the configuration - the image should be resized to the configuration width
 - If the image height is greater than the height specified in the configuration - the image should be resized to the configuration height
 - The image should be resized proportionally
 - Resized version of the image should be saved to the .renditions directory
 - Image path/name should be generated using the same approach as thumbnail generation
(see `\Magento\Cms\Model\Wysiwyg\Images\Storage::resizeFile`)

### Implementation
 - A `\Magento\MediaGalleryRenditionsApi\Api\GenerateRentitionsInterface` should be introduced with a single method `execute(string $path): void`
- An interface should be implemented by `\Magento\MediaGalleryRenditions\Model\GenerateRentitions`
- `GenerateRentitionsInterface` should be used to execute rendition generation during the image upload

### PR DESCRIPTION
**Pull request by konarshankar07**
### Description (*)

This PR will generate renditions image when image is uploaded using Magento Enhanced Media Gallery Image Uploader
### Fixed Issues (if relevant)

1. Fixes magento/adobe-stock-integration#1476: Generate renditions for all images uploaded to magento

### Manual testing scenarios (*)

1.  Configure width/height field from system configuration fields ()
2. Upload image using Enhanced media gallery
3. Check pub/media directory to see the `.renditions` images

### Expected Result
image

image

### CODE REVIEW

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by konarshankar07** [comment]
@magento run all tests

File: MediaGalleryUi/Model/UploadImage.php
Comment (sivaschenko) [inline MediaGalleryUi/Model/UploadImage.php]: To cover all the cases of image upload:
 - Please move the generating renditions to synchronization process. Considering that `SynchronizeFilesInterface` is used here, using the renditions generation will cover both upload from media gallery and `bin/magento media-gallery:sync`
 - Please add renditions generation call to `\Magento\MediaGalleryIntegration\Plugin\SaveImageInformation::afterSave`
 - Please add renditions generation call to `\Magento\MediaGalleryIntegration\Plugin\SaveBaseCategoryImageInformation::afterMoveFileFromTmp`

File: MediaGalleryRenditionsApi/Api/ConfigInterface.php
Comment (sivaschenko) [inline MediaGalleryRenditionsApi/Api/ConfigInterface.php]: The `ConfigInterface` is not a service so it's better to keep it in the `Model` namespace (it is still the API module).

File: MediaGalleryRenditions/Model/GenerateRenditions.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GenerateRenditions.php]: Please do not use the `_` property names prefix

File: MediaGalleryRenditions/Model/RenditionsImageManagement.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/RenditionsImageManagement.php]:
```suggestion
    private $directory;
```

File: MediaGalleryRenditions/Model/RenditionsImageManagement.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/RenditionsImageManagement.php]: Let's create the directory only when creating a rendition. Let's just return the constructed rendition path from this service and do not check if directory or file exists

File: MediaGalleryRenditions/Model/RenditionsImageManagement.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/RenditionsImageManagement.php]: Let's rename the class and add a `Magento\MediaGalleryRenditionsApi\Api\GetRenditionPathInterface`
```suggestion
class GetRenditionPath implements GetRenditionPathInterface
```

File: MediaGalleryRenditions/Model/RenditionsImageManagement.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/RenditionsImageManagement.php]: I think it would be handier to provide ability to get rendition path based on asset id
```suggestion
    public function execute(int $assetId): string
```

File: MediaGalleryRenditionsApi/Api/GenerateRenditionsInterface.php
Comment (sivaschenko) [inline MediaGalleryRenditionsApi/Api/GenerateRenditionsInterface.php]: Considering that there are use cases of assets processing in bulk, I think it would be better to make this service bulk as well
```suggestion
    public function execute(array $paths): void;
```

File: MediaGalleryRenditionsApi/Api/GenerateRenditionsInterface.php
Comment (sivaschenko) [inline MediaGalleryRenditionsApi/Api/GenerateRenditionsInterface.php]: It would be good to ensure only LocalizedException can be thrown from an API service call

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the pull request @konarshankar07 ! Please see my review comments

**Comment by sivaschenko** [comment]
@konarshankar07 as for test coverage: I think it's better to cover the whole functionality by the integration tests for `GetRenditionInterface` and `GenerateRentionsInterface` services. I don't think unit tests are necessary

File: MediaGalleryRenditionsApi/Api/ConfigInterface.php
Comment (konarshankar07) [reply MediaGalleryRenditionsApi/Api/ConfigInterface.php]: Okay, But I moved into the `Api` namespace because of the inconsistency with the other module. If you check the other `ConfigInterface` in the repository they are under the `Api` namespace so should I move the other `ConfigInterface` to the `Model` namespace?

File: MediaGalleryRenditions/Model/GenerateRenditions.php
Comment (konarshankar07) [reply MediaGalleryRenditions/Model/GenerateRenditions.php]: Okay, But may I ask why its not recommended?
Thanks

Comment (sivaschenko) [reply MediaGalleryRenditions/Model/GenerateRenditions.php]: Based on the PSR https://www.php-fig.org/psr/psr-2/#42-properties

File: MediaGalleryRenditionsApi/Api/ConfigInterface.php
Comment (sivaschenko) [reply MediaGalleryRenditionsApi/Api/ConfigInterface.php]: ConfigInterface is not designed to be used outside of the MediaGalleryRenditions module so it's not part of actual module's API, but it still can be customized (preference, plugin, etc)

**Comment by konarshankar07** [comment]
@magento run all tests

File: MediaGalleryRenditions/Model/GenerateRenditions.php
Comment (gabrieldagama) [inline MediaGalleryRenditions/Model/GenerateRenditions.php]:
```suggestion
        if ($this->isResizeable($path)) {
```

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (gabrieldagama) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]: Since this is a private function and we (currently) are not using the $checkFile I would remove it and just not perform the check, or, if the check is important I would always do it. I don't think we need to add this complexity here as an option for a client on private function.

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (gabrieldagama) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]: I think we should negate this condition and throw an exception if true, you can place the rest of the function outside the if. It will make it easier to understand the validation and you will be able to set strict type on function return.

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (gabrieldagama) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]: By having getRenditionsPath throwing an exception we will be able to avoid this if.

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (gabrieldagama) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]:
```suggestion
        return $renditionImageDirectoryPath . DIRECTORY_SEPARATOR . $this->ioFile->getPathInfo($path)['basename'];
```

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (gabrieldagama) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]:
```suggestion
            $renditionPath = $this->getRenditionsRoot();
            if ($relativeFilePath === basename($filePath)) {
                $renditionPath .= DIRECTORY_SEPARATOR;
            }
            $renditionPath .= $relativeFilePath;
```

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (gabrieldagama) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]: Lets make this $filePath strict string and adapt the validation?

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @konarshankar07 thanks for the PR, it look great! Please follow some comments, let me know if you have any questions!

File: MediaGalleryRenditions/Model/RenditionsImageManagement.php
Comment (konarshankar07) [reply MediaGalleryRenditions/Model/RenditionsImageManagement.php]: @sivaschenko ... I think we can get the assets by paths using `Magento\MediaGalleryApi\Api\GetAssetsByPathsInterface` so instead of passing assetId we can pass the `Magento\MediaGalleryApi\Api\Data\AssetInterface` as parameter. Let me know your thoughts

**Comment by sivaschenko** [comment]
Hi @konarshankar07 thanks for the pull request! Would you like to continue progress on it?

**Comment by konarshankar07** [comment]
Hello @sivaschenko ..
Yes, I'm working on this but Can you please give the reply for https://github.com/magento/adobe-stock-integration/pull/1507#discussion_r445981166?
Thanks

File: MediaGalleryRenditions/Model/RenditionsImageManagement.php
Comment (sivaschenko) [reply MediaGalleryRenditions/Model/RenditionsImageManagement.php]: If it's more comfortable for the clients to use the path as an argument - that's ok with me

**Comment by sivaschenko** [comment]
@konarshankar07 replied
Thanks

**Comment by konarshankar07** [comment]
@sivaschenko and @gabrieldagama 
I've finished the feedback changes.
Thanks

File: MediaGalleryIntegration/Plugin/SaveBaseCategoryImageInformation.php
Comment (sivaschenko) [inline MediaGalleryIntegration/Plugin/SaveBaseCategoryImageInformation.php]: please add an empty line between properties

File: MediaGalleryRenditions/Model/GenerateRenditions.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GenerateRenditions.php]: For the consistency
```suggestion
    private function getResizedWidth(): int
```

File: MediaGalleryRenditions/Model/GenerateRenditions.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GenerateRenditions.php]:
```suggestion
    private function getResizedHeight(): int
```

File: MediaGalleryRenditions/Model/GenerateRenditions.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GenerateRenditions.php]:
```suggestion
    private function isResizeable(int $height, int $width): bool
```

File: MediaGalleryRenditions/Model/GenerateRenditions.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GenerateRenditions.php]: Should be an array of `paths` to comply with the interface.

However, returning to the discussion we had, it looks like all the places calling this method have access to an `AssetInterface` and this method implementation requires an `AssetInterface` as well, to avoid double `createAssetFromFile` call, wouldn't it be better to accept an `AssetInterface`  as a parameter?

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]: Write access is not required
```suggestion
        $this->directory = $filesystem->getDirectoryRead(DirectoryList::MEDIA);
```

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GetRenditionPath.php]: I wonder if we can simplify this method implementation to
```suggestion
        if (!$this->directory->isFile($asset->getPath())) {
             throw new InvalidArgumentExcetpion(__('Incorrect asset path!'))
        }
        if (!$this->isResizable($asset)) {
             return $asset->getPath();
        }
        return $this->directory->getRelativePath(RENDITIONS_DIRECTORY_NAME . '/' . $this->directory->getRelativePath($asset->getPath()));
```
The first condition should ensure that asset path is within media directory

File: MediaGalleryIntegration/Plugin/SaveImageInformation.php
Comment (sivaschenko) [inline MediaGalleryIntegration/Plugin/SaveImageInformation.php]: @konarshankar07 @gabrieldagama I would prefer to use `/` instead of `DIRECTORY_SEPARATOR`.

Do you see any possible issues with this approach?

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @konarshankar07 , very nice implementation, please see my comments.

Also, can you please cover the API interfaces with integration tests

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (konarshankar07) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: Write access is required to create the rendition directory path

File: MediaGalleryIntegration/Plugin/SaveImageInformation.php
Comment (konarshankar07) [reply MediaGalleryIntegration/Plugin/SaveImageInformation.php]: I think @gabrieldagama prefer to use `DIRECTORY_SEPARATOR` because of  Windows and Linux have different directory separator but Magento is not compatible with Windows so I think both the approaches is fine for me

**Comment by konarshankar07** [comment]
@sivaschenko ... If you approve these changes then I will work on integration testing
Thanks

File: MediaGalleryIntegration/Plugin/SaveImageInformation.php
Comment (gabrieldagama) [reply MediaGalleryIntegration/Plugin/SaveImageInformation.php]: I usually go for this constant because I find it easier to read, but '/' is fine as well.

Comment (sivaschenko) [reply MediaGalleryIntegration/Plugin/SaveImageInformation.php]: I think '/' should be fine for Windows if we are constructing paths. Using the `DIRECTORY_SEPARATOR` constant can result into issues where '\\' is used in URL links (considering that Magento filesystem can be switched to https adapter).

So functionally I think using `DIRECTORY_SEPARATOR` is more risky for Windows OS comparing to '/' in most cases.

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (sivaschenko) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: This class is responsible for constructing and returning a rendition path only, I don't think it should not create directories

**Comment by sivaschenko** [comment]
@konarshankar07 there are comments that have not been addressed, can you please check them out?

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (konarshankar07) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: I don't think we need to return the asset path if the image size is not resizeable cause GetRenditionPath should be called if the image is resizeable. Let me know your thoughts
Thanks

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (konarshankar07) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: So how we will handle the issue if the rendition directory is not available in the media directory?

**Comment by konarshankar07** [comment]
@sivaschenko ... I have replied to your comments
Thanks

File: MediaGalleryIntegration/Plugin/SaveImageInformation.php
Comment (konarshankar07) [reply MediaGalleryIntegration/Plugin/SaveImageInformation.php]: So Do you think we need to change all `DIRECTORY_SEPARATOR` to '/'?

Comment (sivaschenko) [reply MediaGalleryIntegration/Plugin/SaveImageInformation.php]: It's not necessary for now, just shared my thoughts to bring the attention, I believe it requires a discussion to take a final decision and reflect it in our standards

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (sivaschenko) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: I think returning the asset path if image is not resizable might simplify the client implementation

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (sivaschenko) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: My point is that `GetRenditionPath` path class does not create a directory anyways, so the write access is not needed. It will be needed during the operation of saving the rendition to the filesystem

**Comment by sivaschenko** [comment]
@konarshankar07 thank you! Please see my responses

File: MediaGalleryRenditions/Model/GetRenditionPath.php
Comment (konarshankar07) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: Sorry, But I'm not able to understand how it simplify the implementation because as per the service name which is `GetRenditionPath` and it should return the rendition path but we are returning the asset path if the image is not resizeable which doesn't make sense to me. Another point if we return the asset path then we need to add an extra condition in the GenerateRenditions service to check if the asset path is not same as generate rendition path before image resizing operation and I think it will make the implementation complex. let me know your thoughts

Comment (sivaschenko) [reply MediaGalleryRenditions/Model/GetRenditionPath.php]: I believe it will be comfortable to use `GetRenditionPath` service without additional checks and conditions. The call to the service should return the path to the image that is ready to be used for the content (either from `.rendition` directory or not).
There is a check for `isResizeable` in the `GenerateRenditions` and I don't think any additional checks are needed.

**Comment by konarshankar07** [comment]
@sivaschenko and @gabrieldagama .. Please review this PR
Thanks

**Review by sivaschenko — CHANGES_REQUESTED** [review]
@konarshankar07 can you please update the branch and resolve conflicts

File: MediaGallerySynchronization/etc/di.xml
Comment (sivaschenko) [inline MediaGallerySynchronization/etc/di.xml]: Can you please move this configuration to renditions module

File: MediaGallerySynchronization/composer.json
Comment (sivaschenko) [inline MediaGallerySynchronization/composer.json]: The dependency can be reveresed by extending configuration from renditions module

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks  for updates @konarshankar07 ! Please see my comments

File: MediaGallerySynchronization/etc/di.xml
Comment (sivaschenko) [reply MediaGallerySynchronization/etc/di.xml]: Also, please use integer array key as it will be used as a sort order

File: MediaGalleryRenditions/Model/GenerateRenditionImages.php
Comment (sivaschenko) [inline MediaGalleryRenditions/Model/GenerateRenditionImages.php]: I'd like to avoid the dependency on non-api module. Looks like to achieve this it's needed to allow generating renditions based on paths (to remove the need to create asset object)

File: MediaGalleryRenditions/etc/module.xml
Comment (sivaschenko) [inline MediaGalleryRenditions/etc/module.xml]: The sequece is not needed, it should not matter which of these modules is loaded first

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for updates @konarshankar07 , please see my review comments.

Also, there is a huge update in 2.0-develop branch, those updates do affect this pull request, however I believe there is no critical conflicts. Could you please update the branch and recheck the functionality to be sure.

File: MediaGalleryRenditions/etc/module.xml
Comment (konarshankar07) [reply MediaGalleryRenditions/etc/module.xml]: If I'm not adding the sequence then `ImportFilesComposite` is not executing the `GenerateRenditionImages`

**Comment by sivaschenko** [comment]
Integration test coverage extracted to https://github.com/magento/adobe-stock-integration/issues/1751

**Comment by sivaschenko** [comment]
@magento run all tests

**Review by sivaschenko — APPROVED** [review]

**Pull request merged**
